File: fetch_anilist/fetch_anilist.rb
require "json"
require "httparty"

SHOWS_QUERY = File.read "queries/shows.gql"
CHARS_QUERY = File.read "queries/show_characters.gql"
SINGLE_SHOW_QUERY = File.read "queries/single_show.gql"
MODIFICATIONS = JSON.parse File.read "edited_shows.json"

def anilist_request(query, variables = {})
  # Make the request
  res = HTTParty.post "https://graphql.anilist.co", headers: {
    "Content-Type" => "application/json",
    "Accept" => "application/json"
  }, body: {
    query: query,
    variables: variables
  }.to_json

  # Handle ratelimits for future requests
  if res.headers["retry-after"]
    puts "   Ratelimited, retrying"
    sleep res.headers["retry-after"].to_i + 1
    anilist_request query, variables
  else
    res
  end
end

@shows = []
@opedOnly = []
@characters = []
@vas = []

def get_shows(page: 1)
  puts "Shows page #{page}"
  res = anilist_request SHOWS_QUERY, page: page
  data = JSON.parse(res.body)["data"]["Page"]
  page_info = data["pageInfo"]
  media = data["media"]
  total = page_info["total"]

  puts "Found #{total} shows."
  p page_info
  media.each.with_index 1 do |show, i|
    puts "#{page}-#{i} Show #{show["id"]}"
    if MODIFICATIONS["removed"].include? show["id"]
      puts " Skipping, inelligible"
      next
    end
    movie = MODIFICATIONS["movies"].include? show["id"]
    puts " Updating format to MOVIE (manual override)" if movie
    @shows.push({
      id: show["id"],
      mal: show["idMal"],
      terms: show["title"].values.concat(show["synonyms"]),
      img: show["coverImage"]["medium"],
      format: movie ? "MOVIE" : show["format"],
      short: MODIFICATIONS["shorts"].include?(show["id"]),
      original: MODIFICATIONS["originals"].include?(show["id"])
    })
    get_show_characters show["id"] if show["characters"]["edges"].size > 0
  end

  get_shows(page: page + 1) if page_info["hasNextPage"]
end

def get_show_characters(show_id, page: 1)
  res = anilist_request CHARS_QUERY, page: page, showId: show_id
  data = JSON.parse(res.body)["data"]["Media"]["characters"]
  page_info = data["pageInfo"]
  edges = data["edges"]

  edges.each do |edge|
    character = edge["node"]
    i = @characters.index { |c| c[:id] == character["id"]}
    unless i.nil?
      @characters[i][:show_ids].push show_id
      next
    end

    english_name = [
      character["name"]["first"],
      character["name"]["last"]
    ].reject(&:nil?).join " "
    @characters.push({
      id: character["id"],
      show_ids: [show_id],
      terms: [
        english_name,
        character["name"]["native"],
        *character["name"]["alternative"]
      ].select {|s| s},
      img: character["image"]["medium"]
    })

    vas = edge["voiceActors"]
    vas.each do |va|
      puts " VA #{va["id"]} for character #{character["id"]}"
      english_va_name = [
        va["name"]["first"],
        va["name"]["last"]
      ].reject(&:nil?).join " "
      @vas.push({
        id: va["id"],
        name: english_va_name,
        image: va["image"]["medium"],
        show: show_id,
        character: character["id"]
      })
    end
  end

  get_show_characters show_id, page: page + 1 if page_info["hasNextPage"]
end

def get_extra_show(show_id)
  res = anilist_request SINGLE_SHOW_QUERY, showId: show_id
  show = JSON.parse(res.body)["data"]["Media"]

  puts "Found show #{show_id}."
  hash = {
    id: show_id,
    mal: show["idMal"],
    terms: show["title"].values.concat(show["synonyms"]),
    img: show["coverImage"]["medium"],
    format: "TV",
    short: false,
    original: MODIFICATIONS["originals"].include?(show["id"])
  }
  @shows.push hash
  get_show_characters show_id if show["characters"]["edges"].size > 0
end

puts "Starting"
start = Time.now
puts "Getting additional shows"
MODIFICATIONS["added"].each do |show_id|
  get_extra_show show_id
end
time = Time.now - start
puts "Finished in #{time}ms"
File.write "../public/data/test3.json", {
  shows: @shows.uniq {|s| s[:id]},
  characters: @characters.uniq {|s| s[:id]},
  vas: @vas.uniq {|s| "#{s[:id]}-#{s[:show]}-#{s[:character]}"},
  opedOnly: @opedOnly.uniq {|s| s[:id]}
}.to_json

File: fetch_anilist/fetch_anilist.js
const fs = require('fs');
const superagent = require('superagent');

const showsQuery = fs.readFileSync('queries/shows.gql', 'utf-8');
const showCharactersQuery = fs.readFileSync('queries/showCharacters.gql', 'utf-8');

const allShows = [];
const allCharacters = [];

function sleep (ms) {
	return new Promise(resolve => {
		setTimeout(resolve, ms);
	});
}

async function anilistRequest (query, variables = {}) {
	try {
		const res = await superagent.post('https://graphql.anilist.co')
			.type('application/json')
			.accept('application/json')
			.ok(() => true) // Don't reject on 4xx/5xx headers
			.send({query, variables});
		console.log(`Made request, ${res.header['x-ratelimit-remaining']} left in the next ${parseInt(res.header['x-ratelimit-reset'], 10) - Math.floor(Date.now() / 1000)} seconds`);
		const retryAfter = res.header['retry-after'];
		if (retryAfter) {
			console.log(`=== Ratelimited, retrying in ${retryAfter} seconds ===`);
			await sleep(retryAfter * 1000);
			return await anilistRequest(query, variables);
		}
		return res;
	} catch (res) {
		console.log(res.status, res);
		process.exit(1);
	}
}

async function getShows (page = 1) {
	const res = await anilistRequest(showsQuery, {page});
	const {pageInfo, media} = res.body.data.Page;
	console.log(`Got shows page ${page} (${Math.min(pageInfo.perPage * page, pageInfo.total)} of ${pageInfo.total})`);

	// Loop over all the shows we got...
	for (const show of media) {
		// Push each show's information to the database
		allShows.push({
			id: show.id,
			terms: [
				show.title.romaji,
				show.title.english,
				show.title.native,
				...show.synonyms
			].filter(s => s),
			img: show.coverImage.medium,
			format: show.format
		});
		// Start getting characters from that show
		await getShowCharacters(show.id);
	}

	// If there's more shows, get more shows
	if (pageInfo.hasNextPage) {
		await getShows(page + 1);
	}
}

async function getShowCharacters (showId, page = 1) {
	const res = await anilistRequest(showCharactersQuery, {showId, page});
	const {pageInfo, nodes} = res.body.data.Media.characters;
	console.log(`Got characters page ${page} for show ${showId} (${Math.min(pageInfo.perPage * page, pageInfo.total)} of ${pageInfo.total})`);

	// For all the characters we just got, push their info to the characters array
	for (const character of nodes) {
		// If we've already registered this character, just add a second show for them
		const index = allCharacters.findIndex(existingChar => existingChar.id === character.id);
		if (index !== -1) {
			allCharacters[index].showIds.push(showId);
			continue;
		}
		// Otherwise, construct a new character thingy
		const englishName = [
			character.name.first,
			character.name.last
		].filter(s => s).join(' ');
		allCharacters.push({
			showIds: [showId],
			id: character.id,
			terms: [
				englishName,
				character.name.native,
				...character.name.alternative
			].filter(s => s),
			img: character.image.medium
		});
	}

	// If there's more characters for this show, get them too
	if (pageInfo.hasNextPage) {
		await getShowCharacters(showId, page + 1);
	}
}

const start = Date.now();
getShows().then(() => {
	const finished = Date.now() - start;
	console.log(`Finished scraping in ${finished} ms`);
	fs.writeFileSync('../public/data/test.json', JSON.stringify({
		shows: allShows,
		characters: allCharacters
	}, null, '\t'), 'utf-8');
	console.log('Done');
	process.exit(0);
});
